File: 20220924_Saturday/20220924.go
// Leetcode: 113. Path Sum II.
// https://leetcode.com/problems/path-sum-ii/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 8 ms, faster than 50.53% of Go online submissions for Path Sum II.
// Memory Usage: 4.7 MB, less than 55.16% of Go online submissions for Path Sum II.
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import "fmt"
func pathSum(root *TreeNode, targetSum int) [][]int {
  ans := [][]int{}
  path := []int{} 
  backtrack(root, 0, targetSum, path, &ans)
  return ans
}

func backtrack(node *TreeNode, sum int, target int, path []int, ans *[][]int){
  if nil == node {
    return
  }
  sum += node.Val
  path = append(path,node.Val)
  //fmt.Printf("path %v\n",path)
  //fmt.Printf("Sum %v Target %v\n", sum, target)

  if sum == target && nil == node.Left && nil == node.Right {
    //fmt.Printf("Sum == Target\n")
    copy1 := make([]int, len(path))
    for i, val := range(path){
      copy1[i] = val
    }
    //fmt.Printf("Copy!: %v\n", copy1)

    *ans = append(*ans, copy1)
  }
  backtrack(node.Left, sum, target, path, ans)
  backtrack(node.Right, sum, target, path, ans)
  path =  path[:len(path)-1]
}

File: 20220924_Saturday/20220924.rb
# Leetcode: 113. Path Sum II.
# https://leetcode.com/problems/path-sum-ii/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 174 ms, faster than 25.58% of Ruby online submissions for Path Sum II.
# Memory Usage: 211.4 MB, less than 88.37% of Ruby online submissions for Path Sum II.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Integer[][]}
def path_sum(root, target_sum)
  @answer, @target = [], target_sum
  @path, @sum = [],0
  backtracking(root)
  @answer
end

def backtracking(node)
  return if node.nil?
  @path.push(node.val)
  @sum += node.val
  if @sum == @target && node.left.nil? && node.right.nil?
    @answer.push(@path.clone)
  end
  backtracking(node.left)
  backtracking(node.right)
  @path.pop
  @sum -= node.val
end
